File: js/graph_bfs.js
let { Graph, Node } = require("./graph.js");

let Graph_Searching_BFS = class extends Graph {
  constructor(graph_obj) {
    super();
    for (let graph_id in graph_obj.nodes) {
      graph_obj.nodes[graph_id].visited = false;
      graph_obj.nodes[graph_id].color = "white";
      graph_obj.nodes[graph_id].time = 0;
      graph_obj.nodes[graph_id].predecessor = Infinity;
    }
    this.graph_obj = graph_obj;
    this.traverse_path = [];
  }

  find_the_node(start_id, end_id) {
    let search_node = function (graph_obj) {
      while (traverseWay.length) {
        let curVal = traverseWay.shift();
        let curNode = curVal[0];
        let curpath = curVal[1];
        traverse_path.push(curNode.label);
        curNode.visited = true;
        curNode.color = "black";
        if (curNode.id === end_id) {
          return curpath;
        } else {
          let neigh = curNode.edge;
          while (neigh) {
            let temp = [...curpath];
            let thenode = graph_obj.nodes[neigh.id];
            if (!thenode.visited) {
              thenode.color = "grey";
              thenode.time = curNode.time + 1;
              thenode.predecessor = curNode.id;
              temp.push(thenode);
              traverseWay.push([thenode, temp]);
            }
            neigh = neigh.next;
          }
        }
      }
    };
    let curNode = this.graph_obj.nodes[start_id];
    let traverseWay = [[curNode, [curNode]]];
    let traverse_path = this.traverse_path;
    let pathNode = search_node(this.graph_obj);
    console.log(traverse_path);
    return pathNode;
  }
};

module.exports = Graph_Searching_BFS;

File: py/graph_bfs.py
from graph import Graph
import math

class Graph_Searching_BFS(Graph):
    def __init__(self, graph_obj: Graph):
        super().__init__()
        for node in graph_obj.nodes.values():
            node.visited = False
            node.color = "white"
            node.time = 0
            node.predecessor = math.inf
        self.graph_obj = graph_obj
        self.traverse_path = []

    def find_the_node(self, start_id: int, end_id: int) -> list:
        def search_node() -> None:
            while traverseWay:
                curVal = traverseWay.pop(0)
                curNode = curVal[0]
                curPath = curVal[1]
                curNode.visited = True
                curNode.color = "black"
                self.traverse_path.append(curNode.label)
                if curNode.id == end_id:
                    return curPath
                else:
                    neigh = curNode.edge
                    while neigh:
                        temp = curPath
                        thenode = self.graph_obj.nodes[neigh.id]
                        if not thenode.visited:
                            thenode.color = "grey"
                            thenode.time = curNode.time + 1
                            thenode.predecessor = curNode.id
                            temp.append(thenode)
                            traverseWay.append([thenode, temp])
                        neigh = neigh.next
                
        curNode = self.graph_obj.nodes[start_id]
        traverseWay = [[curNode, [curNode]]]
        pathNode = search_node()
        print(self.traverse_path)
        return pathNode
